feat: Add eth_sign friction

NicolasMassart opened the pull request:
<!--
Thanks for your contribution!

Please ensure that any applicable requirements below are satisfied before submitting this pull request. This will help ensure a quick and efficient review cycle.
-->

## Development & PR Process
1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add `needs-qa` label when dev review is completed
5. Add `QA Passed` label when QA has signed off

## Description

_1. What is the reason for the change?_

Add friction when enabling the eth_sign RPC endpoint

_2. What is the improvement/solution?_

- Update the setting switch to clarify the risk
- Add a bottom sheet that pops when attempting to switch the value to on
- have a two steps friction on the bottom sheet: 1 - checkbox, 2 - text field to fill (case insensitive, like for "delete wallet" and prevent pasting)
- Enable setting only after all friction steps completed
- add tracking events for analytics

## QA builds

- [iOS install page](https://app.bitrise.io/app/be69d4368ee7e86d/build/850fcc6d-74cf-4c8f-b549-a204c7be715b/artifact/1c2d29c97af09c58/p/7038e0571cc39beea13c7ffeec0cbd7a)
- [Android install page](https://app.bitrise.io/app/be69d4368ee7e86d/build/6e0c1307-3e64-48d9-9cb3-a09f0e7b0cc8/artifact/870a158c4fa9f663/p/508e102029406c1747ec4410bb05301b)

[Other QA Artifacts are available on Bitrise Pipeline build #8122](https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/01bf29f9-c91b-4fef-b40f-144f7200f6b1?tab=artifacts)

## Screenshots/Recordings

### Light / Dark modes (on iOS)

| Mode | Advanced setting (off) | CheckBox step | TextInput step (matching) | Advanced setting (on) |
|---|---|---|---|---|
| Light Mode | ![Light mode Advanced setting (off)](https://github.com/MetaMask/metamask-mobile/assets/4677568/304a035e-b3fc-4b02-8df2-5534d2ff4e3d)| ![Light mode Checkbox step](https://github.com/MetaMask/metamask-mobile/assets/4677568/17293fa0-f775-4adf-982d-e88ae814cd2f) | ![Light mode Text field step](https://github.com/MetaMask/metamask-mobile/assets/4677568/d4dc8b26-0aad-4288-bcd1-e3140726494a) | ![Light mode Advanced setting (on)](https://github.com/MetaMask/metamask-mobile/assets/4677568/17145b39-12dc-438b-af03-fe32caebe6f8) |
| Dark Mode | ![Dark Mode Advanced setting (off)](https://github.com/MetaMask/metamask-mobile/assets/4677568/34f37142-28fe-4a87-ba5a-6f579953b739) | ![Dark Mode Checkbox step](https://github.com/MetaMask/metamask-mobile/assets/4677568/8d6578b5-97cd-43f0-98e4-462c22676e79) | ![Dark Mode Text field step](https://github.com/MetaMask/metamask-mobile/assets/4677568/c2bb29f5-bc8d-4945-a4eb-50fdbf008d58) | ![Mode Text Advanced setting (on)](https://github.com/MetaMask/metamask-mobile/assets/4677568/67af4914-1f50-412a-bb88-5c018afa3deb) |

### Screen recording (Light mode on iOS)

https://github.com/MetaMask/metamask-mobile/assets/4677568/ec19f75d-1d3a-46dc-b4dc-43ec4e1bcf9f

## Metrics

metrics event log look like this:

```
 LOG  [MetaMask DEBUG]: Analytics 'trackEvent' - {"category": "Settings Viewed", "properties": {"action": "eth_sign_checkbox_seen"}} {} undefined undefined
 LOG  [MetaMask DEBUG]: Analytics 'trackEvent' - {"category": "Settings Viewed", "properties": {"action": "eth_sign_input_seen"}} {} undefined undefined
 LOG  [MetaMask DEBUG]: Analytics 'trackEvent' - {"category": "Settings Updated", "properties": {"action": "eth_sign_enabled"}} {} undefined undefined
 LOG  [MetaMask DEBUG]: Analytics 'trackEvent' - {"category": "Settings Updated", "properties": {"action": "eth_sign_disabled"}} {} undefined undefined
```
Lines meaning:

1. switch was off and I clicked on i, popping the bottom sheet, viewing the checkbox
2. clicked on continue and moved to second sheet step with text input
3. clicked enabled, it sets the eth_sign setting value to true and closes the bottom sheet
4. then clicked switch off again in the advanced settings without the bottom sheet shown.

looks like the following on MixPanel:
<img width="1488" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/aa544ba3-216b-43fd-8b94-108fa0f94b1f">
![image](https://github.com/MetaMask/metamask-mobile/assets/4677568/6b7efa3b-ba9b-4759-a9a4-456e2ab33eef)

## Issue

fixes MetaMask/mobile-planning#918

## Checklist

* [x] There is a related GitHub issue
* [x] Tests are included if applicable
* [x] Any added code is fully documented

NicolasMassart committed:
initial setting and sheet work

NicolasMassart committed:
reapply text and style changes

NicolasMassart committed:
fix bottom sheet position when keyboard rises

NicolasMassart committed:
format

NicolasMassart committed:
remove tests from pasted component

actual tests to be readded

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

# Conflicts:
#	app/components/Views/Settings/AdvancedSettings/__snapshots__/index.test.tsx.snap
#	locales/languages/en.json

NicolasMassart committed:
fix string keys

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart committed:
fix text

NicolasMassart committed:
prevent paste in text field

NicolasMassart committed:
sheetBottom test

NicolasMassart committed:
advanced settings eth_sign switch test

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart committed:
revert format on wdio side

NicolasMassart committed:
fix AdvancedSettings unit test

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart committed:
WIP using components lib

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart committed:
testing component

fail as the readMore is not implemeted yet

NicolasMassart committed:
implement read more link and test

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart committed:
Merge branch '918_add_eth_sign_setting_friction' of github.com:MetaMask/metamask-mobile into 918_add_eth_sign_setting_friction

NicolasMassart committed:
add the warning message in advanced settings

this requires translation

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart committed:
Merge branch '918_add_eth_sign_setting_friction' of github.com:MetaMask/metamask-mobile into 918_add_eth_sign_setting_friction

NicolasMassart committed:
update button to use isDisabled prop

and snapshots

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart committed:
Merge branch '918_add_eth_sign_setting_friction' of github.com:MetaMask/metamask-mobile into 918_add_eth_sign_setting_friction

NicolasMassart committed:
update snapshot because of bottomSheet now using KeyboardAvoidingView

NicolasMassart committed:
add metrics

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart committed:
revert unrelated format fix

NicolasMassart committed:
styles cleanup

NicolasMassart committed:
fix code formatting and add comments

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart committed:
update text field and buttons style to match design

NicolasMassart committed:
Merge branch '918_add_eth_sign_setting_friction' of github.com:MetaMask/metamask-mobile into 918_add_eth_sign_setting_friction

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

Cal-L commented on app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:
```diff
@@ -0,0 +1,103 @@
+      color: colors.error.default,
+    },
+    warning: {
+      ...fontStyles.normal,
```
Move typography styles to `warningText` and replace with `sBodyMD` instead of hardcoded values

Cal-L commented on app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:
```diff
@@ -0,0 +1,103 @@
+      flex: 1,
+    },
+    bold: {
+      ...fontStyles.bold,
```
Replace with `sBodyMDBold` from typography theme

Cal-L commented on app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:
```diff
@@ -0,0 +1,103 @@
+      flexDirection: 'row',
+      alignItems: 'stretch',
+      justifyContent: 'flex-start',
+      display: 'flex',
```
Remove `display: flex`

Cal-L commented on app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:
```diff
@@ -0,0 +1,103 @@
+      alignItems: 'stretch',
+      justifyContent: 'flex-start',
+      display: 'flex',
+      fontSize: 14,
```
Remove font style from View

Cal-L commented on app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:
```diff
@@ -0,0 +1,103 @@
+      marginEnd: 8,
+    },
+    understandCheckText: {
+      marginEnd: 16,
```
Missing typography `sBodyMD`

Cal-L replied:
+ provide `color` property

NicolasMassart replied:
Why not the default color?

Cal-L commented on app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:
```diff
@@ -0,0 +1,103 @@
+      fontSize: 14,
+    },
+    textConfirmField: {
+      fontSize: 14,
```
Replace font styles with `sBodyMDBold` and provide `color` property

Cal-L commented on app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:
```diff
@@ -0,0 +1,103 @@
+      textAlign: 'left',
+    },
+    textConfirmWarningText: {
+      fontSize: 12,
```
Provide `color` property and use `sBodySM`

Cal-L commented on app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:
```diff
@@ -0,0 +1,103 @@
+    },
+    buttonsContainer: {
+      flexDirection: 'row',
+      display: 'flex',
```
Remove display flex

Cal-L commented on app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:
```diff
@@ -0,0 +1,103 @@
+    },
+    buttonStart: {
+      flex: 1,
+      marginEnd: 8,
```
remove `marginEnd` here and apply `marginHorizontal` to `buttonsContainer`

NicolasMassart replied:
Any idea how I can create the gap between the two buttons without this as we don't have access to the `gap` flex yet in the current RN version?

Cal-L commented on app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:
```diff
@@ -0,0 +1,103 @@
+    },
+    buttonEnd: {
+      flex: 1,
+      marginStart: 8,
```
remove `marginEnd` here and apply `marginHorizontal` to `buttonsContainer`

Cal-L commented on app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:
```diff
@@ -0,0 +1,103 @@
+      width: '100%',
+      paddingVertical: breakPoint ? 16 : 24,
+      justifyContent: 'center',
+      alignSelf: 'center',
```
This section should take up the width completely with horizontal margins and the content should be aligned to the left (by default). `alignSelf: center` is centering it at the moment.

Cal-L commented on app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:
```diff
@@ -0,0 +1,103 @@
+import { fontStyles } from '../../../../../styles/common';
+import Device from '../../../../../util/device';
+
+const breakPoint = Device.getDeviceHeight() < 700;
```
Why is this needed and does it break the UI if it isn't included?

Cal-L commented on app/components/Views/Settings/AdvancedSettings/EthSignFriction/EthSignFriction.styles.ts:
```diff
@@ -0,0 +1,103 @@
+ * @returns StyleSheet object.
+ */
+
+export default (colors: any) =>
```
Type this with `import { ThemeColors } from '@metamask/design-tokens/dist/js/themes/types';`

Cal-L reviewed (commented):
Left some comments

Cal-L commented:
Please also run the app against dark mode since it will show you the missing color properties that need to be applied.

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart commented on app/component-library/components/Sheet/SheetBottom/SheetBottom.tsx:
```diff
@@ -233,7 +235,12 @@ const SheetBottom = forwardRef<SheetBottomRef, SheetBottomProps>(
-      <View style={styles.base} {...props}>
+      <KeyboardAvoidingView
+        behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
+        keyboardVerticalOffset={Platform.OS === 'ios' ? -20 : 64}
```
screenBottomPadding is too much on iOS and not enough on Android. Any idea how to fix this?

NicolasMassart replied:
screenBottomPadding value is 34 on iOS and 0 on Android which is not close to my need.

NicolasMassart committed:
fixed naming, styles and removed use of testId

added accessibility to replace testid

NicolasMassart committed:
Merge branch '918_add_eth_sign_setting_friction' of github.com:MetaMask/metamask-mobile into 918_add_eth_sign_setting_friction

NicolasMassart committed:
fix styles and remove testid

add marginStart on second element of flex (primary + secondary buttons, checkbox + text)
since we don't have flex gap in this RN version yet

NicolasMassart committed:
fix font and color type

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart committed:
Merge branch '918_add_eth_sign_setting_friction' of github.com:MetaMask/metamask-mobile into 918_add_eth_sign_setting_friction

NicolasMassart committed:
update snapshot

NicolasMassart committed:
fix button style

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart committed:
fix styles and bottom sheet

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart committed:
forgot two text colors

NicolasMassart committed:
fix useSafeAreaFrame mocks

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

Cal-L committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

Cal-L reviewed (dismissed):
LGTM

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction

# Conflicts:
#	app/components/Views/Settings/AdvancedSettings/index.js

NicolasMassart committed:
Merge branch '918_add_eth_sign_setting_friction' of github.com:MetaMask/metamask-mobile into 918_add_eth_sign_setting_friction

Cal-L reviewed (approved):
Lgtm

seaona commented:
A couple of findings/comments from QA, the rest looks good :100: 

## Sending ENTER twice triggers other functionalities (i.e. Reset Account function, IPFS Gateway..)
Whenever I input the text and I submit ENTER key, I see that the the first one validates the text, but the second one, instead of `Enabling` the function it triggers another action, depending on where I am on the background screen. This could be Reset Account, Show Hex Data, IPFS Gateway etc..


https://github.com/MetaMask/metamask-mobile/assets/54408225/f8910ad5-5f96-4139-9eb0-038a1ea18710

## `Enable` button is not enabled unless I submit the text
This I am not sure if it's an issue, but commenting just in case. Whenever I enter the correct text, I can see how the button is not enabled, unless I submit (ENTER)


https://github.com/MetaMask/metamask-mobile/assets/54408225/bbe40bf9-7baa-4e48-9dcf-898737aac2a7

## Bold text
I can see how the ON/OFF text is not in bold in Android. Not sure if this is expected

![image](https://github.com/MetaMask/metamask-mobile/assets/54408225/73b6ecb4-cf59-462d-92b9-eaf7dfcc85e9)

NicolasMassart commented:
thanks @seaona, I will investigate the first issue "Sending ENTER twice triggers".

For the second one, I'm going to confirm with design team what the expected behaviour is as this is what I understood from their Figma design (link in the issue)

For the third one, it doesn't look very bold but it is 😅 this is the `sBodyLGMedium` style as indicated in the design.
<img width="458" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/1b80eb0a-6e43-4708-8223-a2c3d526760a">

seaona commented:
thank you for your comments @NicolasMassart @hesterbruikman ! A summary of our discussions and outcomes:
 - **Issue 1**: would be tackled in a separate PR. The issue is here https://github.com/MetaMask/metamask-mobile/issues/6613
 - **Issue 2**: seems that input validation behavior is different in several places, so this is not a blocker and can be consolidated on a separate PR
 - **Issue 3**: not an issue as per @NicolasMassart 's comment
Given that, PR looks good from QA side

NicolasMassart committed:
Merge branch 'main' into 918_add_eth_sign_setting_friction
